### DISCUSSION TITLE
Load .env file

### DISCUSSION THREAD

**Discussion by SimoMay** [nitrojs/nitro#238] [Q&A]
Hi,

I see you using c12 and dotenv (installed but never used in code as far as I can see) but I can't get the .env file to load in config, the only way that work is to add an extends with a ts file

What ways I can load .env file into Nitro?

**Comment by harshagarwalnyu** [comment]
Nitro loads `.env` automatically in development via `dotenv`. In production, `.env` files are intentionally not loaded — environment variables are expected to come from your deployment environment (the host's env vars, not a file).

The recommended pattern:

1. Define your variables in `nitro.config.ts` (or `nuxt.config.ts` for Nuxt):

```ts
// nitro.config.ts
export default defineNitroConfig({
  runtimeConfig: {
    mySecret: '',          // private, server-only
    public: {
      apiBase: '/api',     // accessible client-side too (Nuxt)
    },
  },
});
```

2. Set them in your `.env` file for local development:

```
NITRO_MY_SECRET=super-secret-value
```

3. Access them in handlers:

```ts
const config = useRuntimeConfig();
console.log(config.mySecret); // the value from env
```

The key is the naming convention: `NITRO_` prefix maps to Nitro's private `runtimeConfig`, and `NUXT_PUBLIC_` maps to public config in Nuxt. Nitro uses `c12` for config resolution which handles the `.env` loading in dev via `dotenv` under the hood, but only when `NODE_ENV !== 'production'`.

If you specifically need to load a `.env` file in production (e.g., for local Docker deployments), you can call `dotenv.config()` at the top of your server entry, but the conventional approach is to set env vars in your deployment platform.
